### DISCUSSION TITLE
Add the ability to push multiple gems

### DISCUSSION THREAD

**Discussion by jenshenny** [ruby/rubygems#5574] [Ideas]
Part of the [MFA RFC](https://github.com/rubygems/rfcs/pull/36/files#diff-3d5cc3acc06fe7e9150fdbfc43399c5ad42572c122187774bfc3a4857df524f1R42-R44). 

Adding the ability to push multiple gems on `gem push` reduces friction for gem owners that own and publish many gems at a time. Instead of entering an OTP code for each gem for those who have MFA enabled, an owner can run `gem push` with multiple gems which requires only one OTP code.

A brief proof of concept was explored: https://github.com/Shopify/rubygems/pull/2/files

**Comment by jchestershopify** [comment]
Just a personal opinion: it would be _super_ nice if this was transactional. That is, a push of multiple gems either accepts _all_ the gems or _none_ of the gems.

The rationale comes from watching a Rails team member releasing Rails. Currently they push one gem at a time and there are often failures due to OTP expiry. This means Rails releases can be in an inconsistent state, where some gems have been pushed and some have not. If overlooked, this could break a lot of downstream systems trying to update Rails (or similar gem constellations).

In terms of implementation, I would guess this would mean that the post operation would have a multi-part MIME body so that each .gem is contained in a single operation.

The downside is that this will require a lot more work than just looping over a list repeatedly until things are done.

**Comment by bkuhlmann** [comment]
:bulb: You might want to consider using [Gemsmith](https://www.alchemists.io/projects/gemsmith) since automatic 2FA auth is built in by default. I write about this in [detail here](https://www.alchemists.io/articles/ruby_gems_multi_factor_authentication).

**Comment by hsbt** [comment]
This is better than nothing.

But I have over the 50+ gems and the general publish tool is `rake release` with bundler task each project dir, not push the multiple gems with `gem push`.

**Reply by simi** [reply]
Would it be helpful to provide a way to temporarily  bypass MFA with short-living token? That way you can call something like `gem login` with new parameter, pass MFA and get back token valid for pushing gems in next 5 minutes with no need to pass MFA again.

**Reply by dorianmariecom** [reply]
the same MFA token is valid at least 30 seconds so you have time to push to your gems

**Comment by rubyFeedback** [comment]
This may be a useful change. It could also allow for both "styles", e. g. "continue, even after failure", as well as the other style being "only do it if all gems are pushed", so users could decide on their own.

Personally I don't really need this though. I am very much in the oldschool camp where I upload one gem at a time; takes more time but fewer issues in the long run. (I use ruby scripts to help me push stuff automatically though, such as running kramdown automatically and autogenerating a README.md from a README.gen file, for instance. And various other things.)

I don't use or need MFA though.

**Comment by dorianmariecom** [comment]
I have this bash script:

```bash
#!/bin/bash

set -e

rm -f *.gem
git pull
git push
gem build template-ruby-parser.gemspec
gem build code-ruby-parser.gemspec

for file in *.gem
do
  gem push $file
done

rm *.gem
```

**Comment by larskanis** [comment]
I would really appreciate the ability to `gem push` several gems with a single MFA. I have 2 projects with gem files for a lot of different platforms. This is `pg` with [11 different gem platforms](https://rubygems.org/gems/pg/versions) and `ffi` with [15 gem platforms](https://rubygems.org/gems/ffi/versions). For each new release I have to click the link in the console, click "Authenticate", touch on my security key and wait until the file is uploaded. And this repeats 11 or 15 times for each release. Quite annoying...

**Reply by bkuhlmann** [reply]
💡 Sorry to keep tooting my own horn, but this is handled by Gemsmith ([mentioned above](https://github.com/orgs/rubygems/discussions/5574#discussioncomment-2840361)). Gemsmith solves this by automatically using your U2F key when plugged into your device. Whenever I publish a new gem, I keep the U2F key in my machine and authentication handles automatically (no need to touch the key either). If it helps, I explain this in greater detail [here](https://alchemists.io/articles/ruby_gems_multi_factor_authentication).
